Sources/QuaternionModule/Transformation.swift

```diff
@@ -0,0 +1,487 @@
+    )
+    assert(
+      // TODO: Replace with `approximateEquality()`
+      abs(.sqrt(axis.lengthSquared)-1) < max(.sqrt(axis.lengthSquared), 1)*RealType.ulpOfOne.squareRoot(),
+      "Given axis must be of unit length."
+    )
+
+    self = Quaternion(halfAngle: phase, unitAxis: axis).multiplied(by: length)
+  }
+
+  /// Transforms a vector by this quaternion.
+  ///
+  /// Quaternions are frequently used to represent three-dimensional
+  /// transformations, and thus are used to transform vectors in
+  /// three-dimensional space. The transformation of an arbitrary vector
+  /// by a quaternion is known as an action.
+  ///
+  /// The canonical way of transforming an arbitrary three-dimensional vector
+  /// `v` by a quaternion `q` is given by the following [formula][wiki]
+  ///
+  ///     p' = qpq⁻¹
+  ///
+  /// where `p` is a *pure* quaternion (`real == .zero`) with imaginary part
+  /// equal to vector `v`, and where `p'` is another pure quaternion with
+  /// imaginary part equal to the transformed vector `v'`. The implementation
+  /// uses this formular but boils down to a simpler and faster implementation
+  /// as `p` is known to be pure and `q` is assumed to have unit length – which
+  /// allows for simplification.
+  ///
+  /// - Note: This method assumes this quaternion is of unit length.
+  ///
+  /// Edge cases:
+  /// -
+  /// - For any quaternion `q`, even `.zero` or `.infinity`, if `vector` is
+  /// `.infinity` or `-.infinity` in any of the lanes or all, the returning
+  /// vector is `.infinity` in all lanes:
+  ///   ```
+  ///   SIMD3(-.infinity,0,0) * q == SIMD3(.infinity,.infinity,.infinity)
+  ///   ```
+  /// - For any quaternion `q`, even `.zero` or `.infinity`, if `vector` is
+  /// `.zero`, the returning vector is also `.zero`.
+  ///   ```
+  ///   SIMD3(0,0,0) * q == .zero
+  ///   ```
+  ///
+  /// - Parameter vector: A vector to rotate by this quaternion
+  /// - Returns: The vector rotated by this quaternion
+  ///
+  /// [wiki]: https://en.wikipedia.org/wiki/Quaternions_and_spatial_rotation#Using_quaternion_as_rotations
+  @inlinable
+  public func act(on vector: SIMD3<RealType>) -> SIMD3<RealType> {
+    guard vector.isFinite else { return SIMD3(repeating: .infinity) }
+    guard vector != .zero else { return .zero }
+
+    // The following expression have been split up so the type-checker
+    // can resolve them in a reasonable time.
+    let p1 = vector * (real*real - imaginary.lengthSquared)
+    let p2 = 2 * imaginary * imaginary.dot(vector)
+    let p3 = 2 * real * imaginary.cross(vector)
+    let rotatedVector = p1 + p2 + p3
+
+    // If the rotation computes without over/underflow, everything is fine
+    // and the result is correct. If not, we have to do the computation
+    // carefully and first unscale the vector, rotate it again and then
+    // rescale the vector
+    if rotatedVector.isNormal { return rotatedVector }
```

Comment:
Hi! Is this check correct?
I'm using `quaternion.act(on: (1, 0, 0))`, thus `rotatedVector` is `(1, 0, 0)`, which by this module definition of `isNormal` always fails, leading to infinite recursion.

Reply:
Hi @madbat , thanks so much for reporting the issue!
I am happy to look into possible issues concerning the implementation. Just a heads-up, it might take a few days.
In the meantime, would you mind sharing more insights into how the quaternion is constructed, or more specifically what are the values of the quaternion acting on the vector? Is the quaternion of unit length and/or is the quaternion itself "normal"?

Reply:
Thanks for the quick reply @markuswntr.

Here's my code which leads to the issue:
```
let q = Quaternion(angle: Double.pi/2, axis: SIMD3(0, 1, 0))
let transformed = q.act(on: SIMD3(1, 0, 0))
```

Reply:
Hi @madbat, thanks for sharing the code.
You were correct. The implementation did not properly account for zero values.
I attempt to provide a more elegant implementation before raising a PR, but this may take a while.
For now I provided a quick fix to the issue at: https://github.com/markuswntr/swift-numerics/tree/quaternion/patch-2
It is not heavily tested, but solves your specific issue and should not cause any side-effects.
Feel free to raise further issues with the current/new implementation and/or suggestions!

Reply:
Awesome, thank you @markuswntr!